### PR TITLE
Topology2 doxygen documentation infrastructure

### PR DESCRIPTION
**Pull request by jsarha**
Now this starts to be a little closer where it should be. There is still a lot of comments to be turned into Doxygen documentation, not to mention, documentation to be written for completely undocumented topology code. I will at least convert the existing comments as soon as the generic appearance of this exercise is accepted.

### CODE REVIEW

File: tools/topology/topology2/doc/topology2-filter.py
Comment (lgirdwood) [inline tools/topology/topology2/doc/topology2-filter.py:1]: need license, copyright and also a description of the process.

File: tools/topology/topology2/doc/topology2-filter.py
Comment (lgirdwood) [inline tools/topology/topology2/doc/topology2-filter.py:329]: some big functions here, can we have inline comments for each code block and a comment description of each function. i.e. what does function do.

File: tools/topology/topology2/doc/build/sof.doxygen
Comment (lgirdwood) [inline tools/topology/topology2/doc/build/sof.doxygen]: this will need to be relative.

**Review by lgirdwood — COMMENTED** [review]
Good start and easy to follow the documentation style in the topology conf. Needs a lot more commentary in the python file and description of processing.
@ranj063 pls review

File: tools/topology/topology2/doc/build/sof.doxygen
Comment (jsarha) [reply tools/topology/topology2/doc/build/sof.doxygen]: Oh. That should not be there at all. Its produced by cmake.

**Comment by jsarha** [comment]
Turned this ready for review to gather more attention, but I still intent to turn at least the most relevant pieces of existing comments into Doxygen documentation before this is ready to merge (in any case it hardly is that yet).

If one wants to generate the documentation do following steps:
1. cd tools/topology/topology2/doc
2. mkdir build
3. cd build
4. cmake ..
5.  make
After the last command you should find the html documentation under: sof/tools/topology/topology2/doc/build/doxygen/html

File: tools/topology/topology2/doc/topology2-filter.py
Comment (ranj063) [inline tools/topology/topology2/doc/topology2-filter.py:113]: typo produces

File: tools/topology/topology2/doc/topology2-filter.py
Comment (ranj063) [inline tools/topology/topology2/doc/topology2-filter.py:185]: typo members

**Comment by ranj063** [comment]
> Now this starts to be a little closer where it should be. There is still a lot of comments to be turned into Doxygen documentation, not to mention, documentation to be written for completely undocumented topology code. I will at least convert the existing comments as soon as the generic appearance of this exercise is accepted.

@jsarha thank you for starting this. Looks good to me. Just a few typos to fix

File: tools/topology/topology2/doc/topology2-filter.py
Comment (lgirdwood) [reply tools/topology/topology2/doc/topology2-filter.py:1]: each new source file needs a copyright.

File: tools/topology/topology2/doc/extra-contents/mainpage.doxy
Comment (lgirdwood) [inline tools/topology/topology2/doc/extra-contents/mainpage.doxy:1]: needs copyright.

**Comment by jsarha** [comment]
- rebased on top of latest main
- mainpage.dox
 - add copyright
- topology2-filter.py
 - fix typos in topology2-filter.py comments
 - fix attribute definitions in topology2 snippets
 - Add IncludeByKey contents into topology2 snippets and create basic #if #include structures
 - add copyright
- Turn comments to doxygen documentation in some common files

**Comment by jsarha** [comment]
Some more comment changed to doxygen. Updated the contents page. Its now autogenerated, but unfortunately I could not get cmake to generate it automatically. Any idea how to do it?

File: tools/topology/topology2/doc/CMakeList.txt
Comment (jsarha) [reply tools/topology/topology2/doc/CMakeList.txt:22]: Did not have time to do this yet. Will do it at some point,

File: tools/topology/topology2/doc/topology2-filter.py
Comment (jsarha) [reply tools/topology/topology2/doc/topology2-filter.py:17]: When I started, I did not yet know what I need. I first thought that I can just translate everything as I go through the file, with very little state information. Parts of the translation still works like that, but then there are many places where more state is needed and bigger chucks has to be processed statefully.

If I would start from scratch now, I would probably use some more advanced method, but this is now here and it pretty much does what is needed, so I leave that excercise to a later time.

File: tools/topology/topology2/doc/topology2-filter.py
Comment (jsarha) [reply tools/topology/topology2/doc/topology2-filter.py:66]: Will do.

File: tools/topology/topology2/doc/topology2-filter.py
Comment (marc-hb) [reply tools/topology/topology2/doc/topology2-filter.py:17]: > If I would start from scratch now, I would probably use some more advanced method, but this is now here and it pretty much does what is needed, so I leave that excercise to a later time.

Please add this "technical debt" warning comment at the top of the file so this does not become a time sink for someone else.

**Comment by jsarha** [comment]
- Improve attribute C type selection code
- Turn topology2-filter.py function comments to pydoc
- Add README to instruct how to build the documentation
- Add license, copyright, and brief comment on what topology2-generate-contents.sh does

**Review by lgirdwood — APPROVED** [review]

File: tools/topology/topology2/include/common/data.conf
Comment (ranj063) [inline tools/topology/topology2/include/common/data.conf:10]: looks like a spurious change. If you can drop this commit @jsarha , we're good to go

Comment (jsarha) [reply tools/topology/topology2/include/common/data.conf]: Oh, tuned here and there an then just assumed the spurious change was intentional and made a separate commit of it. Its now fixed.

**Comment by jsarha** [comment]
Some minor typofixes. And dropped an idle commit of spurious change.

**Review by ranj063 — APPROVED** [review]

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:10]:
```suggestion
cat >> extra-contents/contents.doxy << EOF_THING1
    /*! \page doc_contents
    \subsection doc_base_classes Base Classes'

EOF_THING1
```

Comment (marc-hb) [reply tools/topology/topology2/doc/topology2-generate-contents.sh:10]: Etc.

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:25]:
```suggestion
echo "This contents list is automatically generated by $0 $@" >> extra-contents/contents.doxy
```

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:1]: Indent all the code  in a `main()` function

```shell
main()
{

 ...

}

main "$@"
```

https://github.com/thesofproject/sof-test/issues/740

File: tools/topology/topology2/doc/topology2-filter.py
Comment (marc-hb) [reply tools/topology/topology2/doc/topology2-filter.py:17]: Ping.

Also: please run `pylint` and fix at least the scariest warnings (I haven't reviewed the Python code, too long)

File: tools/topology/topology2/doc/README
Comment (marc-hb) [inline tools/topology/topology2/doc/README:6]: This is very convoluted way to use CMake. Replace everything with these two lines:
```suggestion
cmake -S tools/topology/topology2/doc -B build_dir1
cmake --build build_dir1

```

File: tools/topology/topology2/doc/README
Comment (marc-hb) [inline tools/topology/topology2/doc/README:7]: make is typically not available on Windows. `cmake --build` always works.

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:5]:
```suggestion
set -e
```

**Comment by jsarha** [comment]
- some pylint fixes
- clean up topology2-generate-contents.sh
- updated README
- still could not get cmake to run  topology2-generate-contents.sh
  - not even with help of https://github.com/thesofproject/sof/pull/8141#discussion_r1338919275

File: tools/topology/topology2/doc/CMakeList.txt
Comment (marc-hb) [reply tools/topology/topology2/doc/CMakeList.txt:22]: Don't define "clean" targets:
- CMake does this automatically when you use it right
- For "distclean" the only reliable, official and documented CMake way is to delete the entire build directory.

File: tools/topology/topology2/doc/README
Comment (marc-hb) [inline tools/topology/topology2/doc/README:5]: Did you ever try these commands? There's a typo in the file name!

```
CMake Error: The source directory 
"sof/tools/topology/topology2/doc" does not appear to contain CMakeLists.txt.
```

**Review by marc-hb — CHANGES_REQUESTED** [review]
I think it would be totally fine NOT to use CMake _at all_ in the very first PR. But please don't submit CMake code and README "ideas" that have never been run.

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:9]:
```suggestion
cat > "$OUTPUT_DOXY" <<EOF
```

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:19]: - `awk` has `grep` built-in.
- You don't need `split`, `-F` is enough
- use `-print0`

```shell

find .. -name \*.conf -print0 |
  xargs -0 awk -F. "/^Class\.${ctype}\./ "'{ name=$3; ... }'


```

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:22]:
```suggestion
		}' | sort -u >> extra-contents/contents.doxy
```

File: tools/topology/topology2/doc/README
Comment (jsarha) [reply tools/topology/topology2/doc/README:5]: I've run "rm -rf build ; cmake -B build ; cmake --build build" in sof/tools/topology/topology2/doc multiple times last night when testing my changes.

But now I see what my problem, also with cmake rules is. I've been all the time editing CMakeList.txt not CMakeLists.txt... editing it again and retrying, never seeing any change, and finally committing it. While all the time CMakeLists.txt is a left over from some earlier attempt.

I'll try to pull this together and make another attempt at generating contents.doxy with cmake, this time editing the correct file.

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (jsarha) [reply tools/topology/topology2/doc/topology2-generate-contents.sh:19]: Otherwise Ok suggestions except one. Will follow the rest.

> * `awk` has `grep` built-in.
> * You don't need `split`, `-F` is enough

It is not. I need first white-space field separator. Then after I have separated the first field with white-space, then I need to split that with '.' as the separator, to pick the class name from:

Class.Base.<class_name> {

There are of course multiple different ways to extract <class_name> from a line like that, I have picked the first that came to my mind, and it works well enough.

> * use `-print0`
> * new lines are automatically escaped after a pipe `|` (or anywhere the shell expects more), so no need for a backslash there (and yes this is a very good point to insert a new line)
> 
> ```shell
> find .. -name \*.conf -print0 |
>   xargs -0 awk -F. "/^Class\.${ctype}\./ "'{ name=$3; ... }'
> ```

**Comment by jsarha** [comment]
- CMakeList.txt to CMakeLists.txt rename
- cmake now runs topology2-generate-contents.sh automatically
- more cleanups to topology2-generate-contents.sh, most of what was suggested by @marc-hb

File: tools/topology/topology2/doc/README
Comment (marc-hb) [inline tools/topology/topology2/doc/README:5]: Since everything (--options, actions, scripts, tools, failures, CI, verbs, nouns, etc.) is always called the same "build", adding just one trailing slash character makes it obvious that it is a directory in this case.
```suggestion
cmake -B build/
cmake --build build/ -v 
```

This cmake file is not verbose at all (cause it does  nothing, just calls a shell script) so might as well use `-v`.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:30]: CMake automatically creates a "clean" target, try:

```
cmake --build build/ -v -- clean
```

Of course it does not work here because this is not a real CMake file but  just an indirection/obfuscation layer for a bash script.

This `doc-clean` target is also an unnecessary layer indirection. The recommended "superclean" practice with CMake is to manually delete the build directory. No need for any CMake code for that.

From https://cmake.org/cmake/help/book/mastering-cmake/chapter/Getting%20Started.html
> Out-of-source builds make it very easy to maintain a clean source tree and allow quick removal of all of the files generated by a build.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:23]: Don't use COMMAND in `add_custom_target()`. For EACH custom COMMAND you need:
- one `add_custom_command()` where the COMMAND is
- one `add_custom_target()` without any COMMAND.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:17]:
```suggestion
	OUTPUT ${PROJECT_BINARY_DIR}/extra-contents/contents.doxy
```
Never create files outside the build directory, this is confusing both users and CMake.

Please also use `set()` to define a variable and avoid duplication.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:13]: This \@\topdir\@ substitution is the only actual CMake feature used in the entire file, all the rest is just an indirection layer. A simple `sed` or `awk` can do the same.

File: tools/topology/topology2/doc/topology2-filter.py
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-filter.py:24]:
```suggestion
logging.basicConfig(filename="filter_debug.txt",...
```
https://docs.python.org/3/howto/logging.html

(and no this is not "overkill", this module is standard, built-in, fast and dead simple to use)

**Review by marc-hb — CHANGES_REQUESTED** [review]
The two most important things performed by CMake are:

1. A dependency graph to speed up incremental builds (change one file recompile just one file)
2. Operating system-independence: same build code for Linux, Windows and macOS.

As it is now, the current CMakeLists.txt achieves **neither**.

- It does not achieve 2. because it's just an unnecessary indirection layer on top of a shell script that does all the actual work and that will never run on Windows.

- It does not achieve 1. because: 1.1. it does not encode dependencies correctly by lack of CMake expertise (which is unfortunately required) 1.2 This is super difficult to fully achieve with doxygen anyway which deals with a gazillion files; some corners must always be cut (carefully).

So,

  => Embed the \@ topdir \@ substitution in topology2-generate-contents.sh thanks to a simple `sed` or `awk` and drop the CMakeLists.txt file entirely for now. CMake can always be added in a future, smaller PR and it will then be easier to start a brand new `CMakeLists.txt` compared to trying to make sense and "preserve" the features of a hasty and broken one. Trying to get a decent CMakeLists.txt file in this PR is going to take a while and will hold all the rest back for no good reason.

I mean this, just drop the current `CMakeLists.txt` and do everything in the shell script which already does almost everything anyway. Invoke the python script from the shell script. Much faster and simpler.

**Comment by jsarha** [comment]
@marc-hb , you have given me many good suggestions, and following them have been very educating. However, now I am not any more sure what to do next. Sure, I can spend couple of hours inveting a wheel, and writing some script that mangles the correct paths to doxygen config file, then run the topology2 class index generator script, and finally run doxygen to produce the documentation tree. I just want to know for sure that this is what you want, and that this review process is converging to some coclusion. 

The current CMakeLists.txt is not very different from what is used to build the FW source documentation. The only difference is the index generator. The proper dependencies are missing from both, since doxygen builds the documentation from all the sources at once, and so the documentation depends from them all at once too. Just like the generated index depends from all the topology2 sources, and from the generator of course. The topology2-filter.py makes no difference here, doxygen just uses it to open and filter all the sources before processing them (This is not the first of the kind, there are several others for shell scripts [1], matlab, perl, what ever). 

So should I give up on CMakeLists.txt and make another script for building the documenation? Or just move the generated index to bindir and maybe try to indicate the documentation and index dependency to all the topology2 sources?

[1] https://github.com/Anvil/bash-doxygen

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Today the most important thing is that we have some documentation that can be generated by building. 

This PR does that and whilst there are opens around CMake here (and in other places), we need to address these based on priority i.e. solutions needed here for CMake will be needed elsewhere too (but they are not blocking us today, maybe block us tomorrow though...) 

@jsarha can you make the suggested changes by @marc-hb except for usage of awk and sed.

**Comment by marc-hb** [comment]
> Sure, I can spend couple of hours inventing a wheel, and writing some script that mangles the correct paths to doxygen config file, 

That should be one or two lines.

> then run the topology2 class index generator script, 
and finally run doxygen to produce the documentation tree. I just want to know for sure that this is what you want,

> and that this review process is converging to some conclusion.

CMake always tends to be a verbosity and complexity nightmare even for the simplest things and I don't see the CMake part of this PR converging quickly. That's why I'm asking to postpone it to the next PR not to hold this one back.

> So should I give up on CMakeLists.txt and make another script for building the documentation?

Yes please but it does not have to be a different script, I think you can just add it to the existing one. Or it can be a different and _very short_ one. This will be much faster because: 1. it should be short either way 2. shell scripting is less powerful but much simpler than CMake 3. Like most people, you seem comfortable with the former but not the latter.

**Comment by jsarha** [comment]
- use python logging in topology2-filter.py
- separate commands and targets in cmake
- add dependencies to cmake
- update README

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:29]: Move the `>` redirection to the `add_custom_command()` that invokes this. You just need to make sure there's white space on both sides of `>`

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:27]: Move this at the top to be more visible and to make sure it catches everything now and in the future.

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:28]: This is usually at the top too.

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:8]: Pretty sure you can use "$0" here and reserve arguments for if/when they're really needed.

File: tools/topology/topology2/doc/topology2-generate-contents.sh
Comment (marc-hb) [inline tools/topology/topology2/doc/topology2-generate-contents.sh:10]: 3 `echo` -> one more readable EOF heredoc.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:50]: This target is literally longer to type and use than the command it runs. This is not the CMake way (which already generates a "clean" target)

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:28]: Don't quote CMake targets, no one ever does.
```suggestion
add_custom_target(doc-contents
```

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:37]:
```suggestion
	DEPENDS doc-contents
```

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [inline tools/topology/topology2/doc/CMakeLists.txt:39]: Does doxygen reads the *.conf files directly? I doubt it. If not directly then remove this line, having it indirectly in the other target is enough and clearer.

**Review by marc-hb — CHANGES_REQUESTED** [review]
Thanks for making the (tedious, painful) CMake effort! This is starting to look good, small changes requested.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (jsarha) [reply tools/topology/topology2/doc/CMakeLists.txt:39]: Yes it does. The *.conf files are the topology2 files I am documenting with doxygen. Well doxygen does not read the files directly. It runs topology2-filter.py with the file name it wants to open as an argument, and topology2-filter.py reads the file and outputs the converted file - with enough C syntax for doxygen to have some understanging about it - to stdout (well that is what gather it must be doing, I have not read the doxygen sources, and this filter part is very thinly documented).

So yes, the resulting document very much depend on the *.conf files. The contents.doxy does too and the document depends on that too, so in that way the dependency is redundant, but if for some reason in the future we drop the contents page, the document is still depending on the *.conf sources.

**Comment by jsarha** [comment]
Can't finish the job today. I'll go to office and lab now, and I do not have time this evening, but I try to pull this together tomorrow.

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (jsarha) [reply tools/topology/topology2/doc/CMakeLists.txt:50]: Can't call it just "clean":
CMake Error at CMakeLists.txt:49 (add_custom_target):
  The target name "clean" is reserved or not valid for certain CMake
  features, such as generator expressions, and may result in undefined
  behavior.

I'll drop the target all together. it should not be needed for anything, now that the dependencies are handled correctly.

**Comment by jsarha** [comment]
- Idle double guotes dropped
- doc-clean target dropped

File: tools/topology/topology2/doc/CMakeLists.txt
Comment (marc-hb) [reply tools/topology/topology2/doc/CMakeLists.txt:39]: I forgot that `topology2-filter.py` is invoked from doxygen, thanks for the clarification!

**Comment by jsarha** [comment]
Oh, I fogot yesterday the cat << EOF change. No that is added too. I think I've now addressed all the issue pointed by @marc-hb .

**Comment by jsarha** [comment]
Oh, and couple of other contents scripts changes too, and their dependency to CMakeLists.txt. But now, this should be good.

**Review by marc-hb — APPROVED** [review]
Thanks for your patience and hard work, CMake is never fun.

**Pull request merged**
